test/interactive/docker.go

```diff
@@ -0,0 +1,113 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Helper functions for running interactive CLI sessions from Go
+package interactive
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"path"
+)
+
+var defaultDockerCommands []string
+
+func init() {
+	defaultDockerCommands = []string{"docker", "run", "-it", "--rm", "--entrypoint", "bash"}
+}
+
+// Env represents a collection of environment variables and their values
+type Env map[string]string
+
+// Docker is mostly an Command preloaded with arguments which setup Docker for running an image interactively.
+type Docker struct {
+	Command
+}
+
+// PromoteFromEnv pulls the named environment variables from the environment and puts them in the Env.
+// It does not stop on error and returns an error listing all the failed values
+func (e Env) PromoteFromEnv(envVars ...string) error {
+	var err error
+	for _, env := range envVars {
+		v := os.Getenv(env)
+		if v == "" {
+			err = fmt.Errorf("Environment Variables '%s' is not set; %v", env, err)
+		} else {
+			e[env] = v
+		}
+	}
+	return err
+}
+
+// Creates an Docker with default Docker command arguments for running interactively
+func NewDocker() Docker {
+	return Docker{NewCommand(defaultDockerCommands...)}
+}
+
+// AddEnv adds arguments so all the environment variables present in e become part of the docker run's environment
+func (d *Docker) AddEnv(e Env) {
+	for k, v := range e {
+		d.AddArgs("-e", fmt.Sprintf("%s=%s", k, v))
+	}
+}
+
+// AddMount add arguments for the --mount command
+func (d *Docker) AddMount(typeStr, source, target string, optAdditionalArgs ...string) {
+	addl := ""
+	for _, arg := range optAdditionalArgs {
+		addl = addl + "," + arg
```

Comment:
This will be `addl := strings.Join(arg, ",")`

Reply:
the leading `,` is necessary with arguments, which rules out your first example, or necessary that the string is empty when no arguments, ruling out the second example unless we add `if len(optAdditionaArgs) > 0` first. Your call

Reply:
I think below is more understandable.
```
if len(optAdditionaArgs) != 0 {
    addl = "," + strings.Join(optAdditionalArgs, ",")
}
```